File: src/bmsx_cpp/machine/devices/input/sample_edge.cpp
#include "machine/devices/input/sample_edge.h"

#include "input/manager.h"
#include "machine/devices/input/action_table.h"
#include "machine/devices/input/event_fifo.h"
#include "machine/devices/input/sample_latch.h"

namespace bmsx {

InputControllerSampleEdge::InputControllerSampleEdge(Input& input, InputControllerSampleLatch& sampleLatch, InputControllerActionTable& actionTable, InputControllerEventFifo& eventFifo)
	: m_input(input)
	, m_sampleLatch(sampleLatch)
	, m_actionTable(actionTable)
	, m_eventFifo(eventFifo) {}

void InputControllerSampleEdge::onVblankEdge(f64 currentTimeMs, u32 nowCycles) {
	if (!m_sampleLatch.consumeVblankEdge(nowCycles)) {
		return;
	}
	m_input.samplePlayers(currentTimeMs);
	m_actionTable.sampleCommittedActions(m_eventFifo);
}

} // namespace bmsx

File: src/bmsx/machine/devices/input/sample_edge.ts
import type { Input } from '../../../input/manager';
import type { InputControllerActionTable } from './action_table';
import type { InputControllerEventFifo } from './event_fifo';
import type { InputControllerSampleLatch } from './sample_latch';

export class InputControllerSampleEdge {
	public constructor(
		private readonly input: Input,
		private readonly sampleLatch: InputControllerSampleLatch,
		private readonly actionTable: InputControllerActionTable,
		private readonly eventFifo: InputControllerEventFifo,
	) {}

	public onVblankEdge(currentTimeMs: number, nowCycles: number): void {
		if (!this.sampleLatch.consumeVblankEdge(nowCycles)) {
			return;
		}
		this.input.samplePlayers(currentTimeMs);
		this.actionTable.sampleCommittedActions(this.eventFifo);
	}
}
